auth/middleware.py

```diff
@@ -0,0 +1,55 @@
+from fastapi import HTTPException, Request
+from fastapi.responses import RedirectResponse
+from functools import wraps
+from loguru import logger
+from jose import jwt
+
+from auth.utils import get_jwt_secret_key
+
+def login_required(jwt_secret_key=None):
+    """
+    Decorator to require login, with an optional jwt_secret_key parameter.
+    
+    Parameters:
+    - jwt_secret_key: Optional JWT secret key. If not provided, it will be obtained from the utils function.
+    """
+    def decorator(func):
+        @wraps(func)
+        async def wrapper(*args, **kwargs):
+            if kwargs.pop('testing', None):
+                return func(*args, **kwargs)
+            
+            logger_service = logger.bind(service="AuthMiddleware")
+            request = kwargs.get("request")
+
+            if not request or not isinstance(request, Request):
+                raise HTTPException(status_code=400, detail="Request object is missing or invalid")
+
+            token = request.cookies.get("token")
+            if not token:
+                return RedirectResponse(url="/login", status_code=302)
+
+            try:
+                # Use the provided jwt_secret_key, or get from utils if not provided
+                secret_key = jwt_secret_key or get_jwt_secret_key()

```

Comment:
The caller shouldn't provide any sort of secret key. Instead, it would be better to wrap the key and this function in a class (perhaps called TokenMiddleware) so it removes any implicit side effect in the background and the caller can just use the function from the class without having to provide their own secret key (which shouldn't really happen anyway)

Reply:
yes, i'll fix that on different PR for refactor purposes